File: lib/core/models/subscription_status.dart
class SubscriptionStatus {
  const SubscriptionStatus({
    this.provider = 'stripe',
    this.status = 'none',
    this.active = false,
    this.bookmarkNotificationsLimit = 0,
    this.rosterAssistantMonthlyCreditUsd = 0,
  });

  final String provider;
  final String status;
  final bool active;
  final int bookmarkNotificationsLimit;
  final double rosterAssistantMonthlyCreditUsd;

  factory SubscriptionStatus.fromJson(Map<String, dynamic> json) {
    return SubscriptionStatus(
      provider: json['provider']?.toString() ?? 'stripe',
      status: json['status']?.toString() ?? 'none',
      active: json['active'] == true,
      bookmarkNotificationsLimit:
          (json['bookmarkNotificationsLimit'] as num?)?.toInt() ?? 0,
      rosterAssistantMonthlyCreditUsd:
          (json['rosterAssistantMonthlyCreditUsd'] as num?)?.toDouble() ?? 0,
    );
  }
}

File: expo/src/features/subscription/subscription-status.ts
export class SubscriptionStatus {
  constructor(
    readonly provider = 'stripe',
    readonly status = 'none',
    readonly active = false,
    readonly bookmarkNotificationsLimit = 0,
    readonly rosterAssistantMonthlyCreditUsd = 0,
  ) {}

  static fromJson(value: unknown): SubscriptionStatus {
    const json = isRecord(value) ? value : {};
    return new SubscriptionStatus(
      typeof json.provider === 'string' ? json.provider : 'stripe',
      typeof json.status === 'string' ? json.status : 'none',
      json.active === true,
      Math.trunc(numberValue(json.bookmarkNotificationsLimit)),
      numberValue(json.rosterAssistantMonthlyCreditUsd),
    );
  }
}

function numberValue(value: unknown): number {
  return typeof value === 'number' && Number.isFinite(value) ? value : 0;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}
